File: android/sample/app/src/main/java/com/unfydqry/searchsample/Highlight.kt
package com.unfydqry.searchsample

import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.withStyle

/// Bridges the engine's `highlight` markers to a styled [AnnotatedString].
///
/// The host asks the engine to wrap each matching region in [OPEN]/[CLOSE]
/// sentinels; the UI turns that marked text into an [AnnotatedString] whose
/// matched spans stand out in the result list.
object Highlight {
    /// Sentinel markers wrapped around matched regions. C0 control characters,
    /// so they never collide with real (normalized) content.
    const val OPEN = "\u0002" // STX
    const val CLOSE = "\u0003" // ETX

    private val matchStyle = SpanStyle(
        background = Color(0x80FFEB3B), // translucent yellow
        fontWeight = FontWeight.Bold,
    )

    /// Parses text wrapped with [OPEN]/[CLOSE] markers into an [AnnotatedString],
    /// emphasizing the matched spans. An optional [prefix] (e.g. "よみ: ") is
    /// prepended unstyled.
    fun annotated(marked: String, prefix: String = ""): AnnotatedString = buildAnnotatedString {
        if (prefix.isNotEmpty()) append(prefix)
        var run = StringBuilder()
        var inMatch = false

        fun flush() {
            if (run.isEmpty()) return
            if (inMatch) withStyle(matchStyle) { append(run.toString()) } else append(run.toString())
            run = StringBuilder()
        }

        marked.forEach { ch ->
            when (ch) {
                OPEN[0] -> { flush(); inMatch = true }
                CLOSE[0] -> { flush(); inMatch = false }
                else -> run.append(ch)
            }
        }
        flush()
    }
}

File: ios/sample/SearchSample/Highlight.swift
import SwiftUI

/// Bridges the engine's `highlight` markers to a styled `AttributedString`.
///
/// `SearchModel` asks the engine to wrap each matching region in `open`/`close`
/// sentinels; the view turns that marked text into an `AttributedString` whose
/// matched spans stand out in the result list.
enum Highlight {
    /// Sentinel markers wrapped around matched regions. C0 control characters,
    /// so they never collide with real (normalized) content.
    static let open = "\u{2}" // STX
    static let close = "\u{3}" // ETX

    /// Parses text wrapped with `open`/`close` markers into an `AttributedString`,
    /// emphasizing the matched spans.
    static func attributed(_ marked: String) -> AttributedString {
        let openChar = Character(open)
        let closeChar = Character(close)

        var result = AttributedString()
        var span = AttributedString()
        var inMatch = false

        func flush() {
            guard !span.characters.isEmpty else { return }
            if inMatch {
                span.backgroundColor = .yellow.opacity(0.5)
                span.foregroundColor = .primary
                span.inlinePresentationIntent = .stronglyEmphasized
            }
            result.append(span)
            span = AttributedString()
        }

        for ch in marked {
            switch ch {
            case openChar:
                flush()
                inMatch = true
            case closeChar:
                flush()
                inMatch = false
            default:
                span.append(AttributedString(String(ch)))
            }
        }
        flush()
        return result
    }
}
